Feature/152 xp reward contract

ITStar10 opened the pull request:
Closes #152

ITStar10 committed:
[VDA-Verification-Base] V0.2.0

- Separate `VDATRustedSignerManager` from `VDAVerificationContract`
- Update test file extensions

ITStar10 committed:
[VDA-XP-Reward] V0.1.0

- Initial version

ITStar10 committed:
[VDA-DID-Registry] V0.2.1

- Add `isRegistered()` function for `VDAXPReward` contract

ITStar10 committed:
[VDA-Reward-Contract] v0.5.2

- Fixed typos in the comment and `ChangeLog.md` file

ITStar10 committed:
[VDA-XP-Reward] Work in progress

- Added `uniqueId` properpty into the `ClaimInfo` struct
- Contract updated
- Test works for claim informations that has not `uniqueId` properties.

To-do
- Add test case for `uniqueId` property

ITStar10 committed:
[VDA-XP-Reword] Work in progress

- Contract updated
- Test code updated (not completed)

To-do
- Separate test cases

ITStar10 committed:
[VDA-XP-Reward] 0.1.0

- Updates on review

tahpot commented on VDA-DID-Registry/contracts/IVeridaDIDRegistry.sol:
```diff
@@ -23,6 +23,14 @@ interface IVeridaDIDRegistry {
+     * @param didAddress DID address to be checked
+     * @return bool `true` if registered
+     */
+    function isRegistered(address didAddress) external view returns(bool);
```
Couldn't you have just used `lookup()`?

tahpot commented on VDA-DID-Registry/contracts/IVeridaDIDRegistry.sol:
```diff
@@ -23,6 +23,14 @@ interface IVeridaDIDRegistry {
+     * @param didAddress DID address to be checked
+     * @return bool `true` if registered
+     */
+    function isRegistered(address didAddress) external view returns(bool);
```
Any reason you didn't use `lookup()`?

ITStar10 replied:
This `isRegistered()` function consumes less gas fees than the `lookup()` function.
In some contracts, we need to check whether a DIDAddress is registered in this `DIDRegistry` contrct.  In case, it's convenient to use the `isRegistered()` function instead of `lookup()` function.

tahpot replied:
If it's read only, it doesn't use gas though?

ITStar10 replied:
It doesn't cosume gas when it's called by the exteranl (through web3).
But if it's called inside a function that uses gases, it consume the gas for this function call too,
In other word, if this `isRegistered()` function is called by the `claimXPReward()` function of the `XPReward` contract, the caller of the `claimXPReward()` function should pay the gas for this `isRegistered()` function too.

tahpot replied:
Ah... that makes sense.

tahpot commented on VDA-Verification-Base/contracts/VDATrustedSignerManager.sol:
```diff
@@ -0,0 +1,79 @@
+//SPDX-License-Identifier: MIT
```
Can you explain the reason for this refactor?

ITStar10 replied:
I just separated the `VDATrustedSignerManager` from original `VDAVerificationBase` contract.
In some cases, we need a contract that includes feature for trusted signers.
For example, in the `claimXPReward()` function of  `XPReward` contract, we didn't add the `requestSignature` and `requestProof` parameters at first. Here, we need to manage the trusted signers. So I separated the `VDATrustedSignerManager` and made the `XPReward` contract inherits that.
Later, we added the `requestSignature` and `requestProof`. So I made the `XPReward` contract inherits the `VDAVerificationBase`.

By the way, we can create some contracts that inherits the `VDATrustedSignerManager` only later.

tahpot replied:
Ok, noted

tahpot commented on VDA-XP-Reward/.env.example:
```diff
@@ -0,0 +1,17 @@
+# Used to verify contract after deployment
+ETHERSCAN_API_KEY="<Input your etherscan.com api key>"
```
Can we remove etherscan?

ITStar10 replied:
Removed

tahpot commented on VDA-XP-Reward/.env.example:
```diff
@@ -0,0 +1,17 @@
+# Used to verify contract after deployment
+ETHERSCAN_API_KEY="<Input your etherscan.com api key>"
+POLYGONSCAN_API_KEY="<Input your polygonscan.com api key"
```
Add a comment describing why this is needed

ITStar10 replied:
Pls see the line#1.
This is used to verify the contract after deployment.

tahpot commented on VDA-XP-Reward/.env.example:
```diff
@@ -0,0 +1,17 @@
+POLYGONSCAN_API_KEY="<Input your polygonscan.com api key"
+
+# Used to deploy contract
+PRIVATE_KEY="<Input your wallet private key for deploying contract>"
```
Is this a Verida private key?

ITStar10 replied:
No, this is the wallet private key for paying gas fees while deploying the contract.

tahpot replied:
Ok, make it clear this is a polygon private key

tahpot commented on VDA-XP-Reward/.env.example:
```diff
@@ -0,0 +1,17 @@
+PRIVATE_KEY="<Input your wallet private key for deploying contract>"
+
+# Used for contract deployment & verification. Optional
+POLYGON_TESTNET_RPC="<Input your own RPC url for polygon mainet>"
```
Is this TESTNET or MAINNET?

ITStar10 replied:
In fact, this is the RPC of `Mumbai` testnet.
Removed.

tahpot commented on VDA-XP-Reward/.env.example:
```diff
@@ -0,0 +1,17 @@
+POLYGON_MAINNET_RPC="<Input your own RPC url for mumbai net>"
+
+# Defender keys : used for multi-sign support scripts
+DEFENDER_TEAM_API_KEY="<Input your Defender Team API Key>"
```
Are all of these required? If so, please explain why

ITStar10 replied:
The `DEFENDER_...` keys are required to make proposals for following:
- Transfer the ownership to multi-sig wallet after the contract deployment
- Make a proposal for upgrading contract to newer version

As you know, we can't deploy the contract wit the multi-sig wallet. So, at the first, we deploy the contract with EOA account and then transfer the ownerships to the multi-sig wallet.
These `DEFENDER_...` keys are generated when we created the multi-sig wallet.

tahpot replied:
Please add the explanation into the code or the README

ITStar10 replied:
Added explain to the `.env.example` file

tahpot commented on VDA-XP-Reward/README.md:
```diff
@@ -0,0 +1,67 @@
+# Contract Explain
```
`# Verida XP Rewards`

ITStar10 replied:
This `README` file has separated into 2 parts:
- `Contract Explain`
- `Test & Deploy`

In the first part, it explains about the contract. In the second part, it explains how to test and deploy contract.

Should I rename the `Contract Explain` to `Verida XP Rewards`?

tahpot replied:
> Should I rename the Contract Explain to Verida XP Rewards?

Yes pls

ITStar10 replied:
Updated

tahpot commented on VDA-XP-Reward/README.md:
```diff
@@ -0,0 +1,67 @@
+
+The key objectives are:
+
+1. Every month, VDA is deposited into the smart contract as reward tokens
```
Update this to match the updated issue (we changed the spec slightly)

ITStar10 replied:
I just copied this description from the github:
https://github.com/verida/blockchain-contracts/issues/152

Let me know how to update this.

tahpot replied:
That issue has been updated since you copied, so just re copy and paste

ITStar10 replied:
Updated

tahpot commented on VDA-XP-Reward/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## Dependencies
+### Verida contracts
+This contract dependes on the followig Verida contract:
```
Fix typos

ITStar10 replied:
Updated

tahpot commented on VDA-XP-Reward/README.md:
```diff
@@ -0,0 +1,67 @@
+## Dependencies
+### Verida contracts
+This contract dependes on the followig Verida contract:
+- `@verida/vda-verification-contract` : VDA-Verification-Base contract
```
Explain why each of these dependencies are required

ITStar10 replied:
Updated. Pls check again.

tahpot replied:
Thanks.

You can remove `did-registry-contract` now?

ITStar10 replied:
Updated.

tahpot commented on VDA-XP-Reward/contracts/DateTime.sol:
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: MIT
```
Can you reference where this was obtained from?

Is this a standard OZ library?

ITStar10 replied:
No, this is not the standard OZ library. Just refered from the know github source.
And this contract is used to calculate the year and month from the timestamp.

ITStar10 replied:
The original code is here : https://github.com/RollaProject/solidity-datetime/blob/master/contracts/DateTime.sol

I just used only the part related to the year and month.

tahpot replied:
Ok, reference where this code came from, this will be helpful with audit.

ITStar10 replied:
Added reference at the above of library declaration.
```ts
pragma solidity ^0.8.18;

// This is refereneced from following : https://github.com/RollaProject/solidity-datetime/blob/master/contracts/DateTime.sol
library DateTime {
```

tahpot commented on VDA-XP-Reward/contracts/IVDAXPReward.sol:
```diff
@@ -0,0 +1,112 @@
+    /**
+     * @notice Claim information
+     * @param typeId - Proof type. Ex: 'gamer31'
+     * @param uniqueId - Unique Id.
```
Optional, provide example

ITStar10 replied:
Just updated as following:
```
* @param uniqueId - Unique Id. Optional Ex : ''.
```

tahpot commented on VDA-XP-Reward/contracts/IVDAXPReward.sol:
```diff
@@ -0,0 +1,112 @@
+    /**
+     * @notice Withdraw tokens
+     * @dev Only the contract owner allowed
+     * @param to - Recipient address
```
Make it clear this is a wallet address

ITStar10 replied:
Updated.

tahpot commented on VDA-XP-Reward/contracts/IVDAXPReward.sol:
```diff
@@ -0,0 +1,112 @@
+    function getRateDenominator() external view returns(uint32);
+
+    /**
+     * @notice Set the denominator for rate values
```
Please add more detail on what this is used for

ITStar10 replied:
Updated, Pls check again.

tahpot replied:
Thanks.

Is there a default value? If so, specify that here as well

ITStar10 replied:
Updated.

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+     * @param info - Claim information
+     */
+    function _validateProofTime(address didAddress, DateTime._DateTime memory curTime, ClaimInfo calldata info) internal virtual view {
+        if (curTime.year != info.issueYear || curTime.month != (info.issueMonth + 1)) {
```
`curTime.year != info.issueYear`

I think this will return a false positive when current date is Jan 2025 (issuer year will not match current year).

tahpot replied:
Please add failed unit tests that cover this scenario.

ITStar10 replied:
Already, consider all cases. It is checked by the `else if` statement.
Test case was already added:

ITStar10 replied:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/224bfda1-1fa9-4c2d-b27d-5ff4d15fdf4f)

tahpot replied:
You need to test blocktime = december, claim time is january. And claim is accepted. You will need to set the blockcchain blocktime to be December to do this properly.

ITStar10 replied:
Added the test case:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/b154c90f-e62f-49e4-bc7f-d7c8768073c3)

tahpot replied:
@ITStar10 

This new test you have added succeeds, but it should fail.

You set the blocktime to be January 2024, but `CLAIM_GAMER31` has month of April 2024:

```
issueYear: 2024,
issueMonth: 4,
```

ITStar10 replied:
Block time is set as 2025/1 by the following code:
```ts
blockTime.setFullYear(blockTime.getFullYear() + 1);
blockTime.setMonth(0);
await time.increaseTo(blockTime.getTime() / 1000);
```
Here, it increases the year.

And, the issueYear and issueMonth of `CLAIM_GAMER31` is updated in the `checkClaimSuccess()` function.
```ts
const checkClaimSuccess = async (claimInfos: ClaimData[], rewardXPs: number[]) => {
                const claimData = [];
                const [issueYear, issueMonth] = await getProofIssueTime(); //Here get the time that is 1 month before of 
 current blockchain time: 2024/12
                ...

                for (let i = 0; i < rewardXPs.length; i++) {
                    ...

                    const claimInfo: ClaimData = {...claimInfos[i]};
                    claimInfo.issueYear = issueYear;  //  Here set 2024
                    claimInfo.issueMonth = issueMonth; // Here set 12
                    ...

```

The `getProofIssueTime()` function gets the time that is 1 month before of the current blockchain time:

Here attach the result of `Claim in January` test:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/6b5470a9-77f7-4c16-ab6a-10ec269b85f5)

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+     */
+    function _validateProofSigner(address didAddress, address proofSigner, ClaimInfo calldata info) internal virtual view {
+        bool isVerified;
+        uint index;
```
Does solidity default value - `0` for index?

ITStar10 replied:
Correct

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+            if (isClaimedTypeId[rawMsg]) {
+                revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+            }
+            isClaimedTypeId[rawMsg] = true;
```
Should this only be set at the very end if all the tests pass?

ITStar10 replied:
Just not clear yet.
In each for loop, it tests whether the `typeId` is claimed. If claimed, it stops the for loop and revert the error.

tahpot replied:
But, let's say it's a duplicate unique ID (checked later in the code), then it will have already set `isClaimedTypeId` true... however it wasn't actually claimed due to the duplicate unique id.

ITStar10 replied:
In the solidity,for the reverted transactions, all the status are not updated.
In fact, Solidity is Turing complete. Means that it can know the result before running the transaction. So it doesn't run any code for the reverted transaction.

So, there won't be any cases that `typeId` updated to true for duplicated uniqueId.

ITStar10 replied:
Thanks as following:
In the solidity, once the transaction completed successfully, all the state values are updated as the logic.
If the transaction failed, all the state values are not updated. They kept original states.

tahpot replied:
Ok, got it, thanks.

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+                if (isClaimedUniqueId[rawMsg]) {
+                    revert DuplicatedUniqueId(didAddress, infos[i].typeId, infos[i].uniqueId, infos[i].xp);
+                }
+                isClaimedUniqueId[rawMsg] = true;
```
Should this only be set at the very end if all the tests pass?

ITStar10 replied:
Same as above.
Checks the `uniqueId` in each for loop.

tahpot replied:
As above

ITStar10 replied:
Same as above.

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+            }
+            isClaimedTypeId[rawMsg] = true;
+
+            if (bytes(infos[i].uniqueId).length != 0) {
```
Please add more comments explaining the logic of this code.

I don't think the logic is correct. It doesn't appear to be verifying the `infos[i]` signature is valid based on the data provided.

ITStar10 replied:
In each for loop, it checks the `signauture` of the claim information.

First, it checks whether the `xp` value is 0. If so, reverted.
Secod, it checks the validity of proof time.
![image](https://github.com/verida/blockchain-contracts/assets/15656252/6a89cca2-e127-446f-b5c4-1ebc9f72ae75)

ITStar10 replied:
Third, it checks whether the `typeId` of the claim information is claimed in this month:
It check for the `${didAddress}${typeId}${year}${month}`.

ITStar10 replied:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/4428b946-7e81-4725-b467-d51bf6d6f026)

ITStar10 replied:
4th : if the `uniqueId` is not optional, it checks whether this `uniqueId` is claimed before for the specified `typeId`:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/2ba420d3-bc68-4d63-982e-02301d450453)

ITStar10 replied:
5th: If the `uniqueId` is optional, it checks the `signature` is claimed before.
Here, the `signature` is encoded by the `did-typeid-year-month-xp`. So checking the `signature` means the checking the `did-typeId-year-mont-xp` value.
![image](https://github.com/verida/blockchain-contracts/assets/15656252/2360cd93-518d-457c-b4c7-7d193f246ab5)

tahpot replied:
I was wanting your comments in the code itself.

As per my comment below, the `signature` isn't being verified correctly.

ITStar10 replied:
Already explained about the Turing completness in the Solidity

ITStar10 replied:
Added comments.
And the verification steps are updated.

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+                isClaimedSignature[infos[i].signature] = true;
+
+            }
+
+            dataHash = keccak256(rawMsg);
```
This should use `VDAVerificationBase.verifyData()`?

ITStar10 replied:
No, this is not the same as the`verifyData()` function.
```
function verifyData( bytes memory data,  bytes memory signature, bytes memory proof)
```
The `verifyData()` function recovers the `contextSigner` from the `data` and `signature`. And then checks this `${trustedSigner}${strContextSigner}` is valid by the `proof`.

By the way, in the `XPRewardContract`, it check the following:
```
dataHash = keccak256(`${didAddress}${typeId}${uniqueId}${year}${month}${xp}`)
signer = recover(dataHash, claimInfo[i].signature)
```
And checks this `signer` is one of the tursted signers.

tahpot replied:
The code is incorrect then, as the `verifyData()` function should be used. That's the purpose of 

The `infos[i].signature` will be the signature from a trusted context signer, just like our other reward contract. We never use blockchain addresses as signers, we always use context signers.

ITStar10 replied:
`infos[i].signautre` is from the trusted signer, correct? Because, it's as following:
```ts
infos[i].signature = TrustedSigner.sign(`${didAddress}${typeId}${uniqueId}${issueYear}${issueMonth}${xp}`)
```
The result of `recover(infos[i].signature)` is the address of the proof signer.
We should check this recoved address is one of the trusted signers.

tahpot replied:
I see the confusion.

You are assuming the truster signer is signing using the `DID` master key.

It should be that the trusted signer is signing using a context signing key. Hence, you should be using the `verifyData()` method which verifies using the context signing key, not the master DID key.

Please ensure the other reward contract is also using the context signing key.

ITStar10 replied:
Updated to use the `verifyData()` function

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+                }
+                isClaimedUniqueId[rawMsg] = true;
+
+                rawMsg = abi.encodePacked(didAddress, rawMsg, infos[i].issueYear, infos[i].issueMonth, infos[i].xp);
```
This is never used, why is it set here?

ITStar10 replied:
This is used outside the `if` statement:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/7bfb536b-a4d1-463b-ab0b-46571426087a)

tahpot replied:
Oops, I missed that.

tahpot commented on VDA-XP-Reward/hardhat.config.ts:
```diff
@@ -0,0 +1,77 @@
+      chainId: 137,
+      accounts: PRIVATE_KEY !== undefined ? [PRIVATE_KEY] : [],
+    },
+    polygonmumbai: {
```
Remove mumbai support everywhere

ITStar10 replied:
removed

tahpot commented on VDA-XP-Reward/test/didHelper.ts:
```diff
@@ -0,0 +1,38 @@
+import EncryptionUtils from "@verida/encryption-utils"
+import { ethers } from "hardhat"
+
+export const endPoints = [
```
What's the code in this file doing?

I'm confused about the need for `endPoints`.

ITStar10 replied:
In the `XPReward` contract, it checks whether the `didAddress` is registered in the `DIDRegistry` contract.
To register a did in the `DIDRegistry` contract, we need endpoints.
These endpoints are used to register a DID in the `DIDRegistry` contract.

tahpot replied:
I assume this will now be removed as we are removing DIDRegistry from this contract.

ITStar10 replied:
Removed all unnecessary codes related to the `DIDRegistry` contract.

tahpot commented on VDA-XP-Reward/test/index.test.ts:
```diff
@@ -0,0 +1,844 @@
+        await deployContracts()
+    })
+
+    describe("TrustedSigners", () => {
```
We can remove these tests as they already existing in the VDAVerificationBase library

ITStar10 replied:
Though it was tested in the `VDAVerificationBase`, it might better to double check this `XPReward` contract support that.
Would you like to remove this code?

tahpot replied:
Ok, let's leave then.

tahpot commented on VDA-XP-Reward/test/index.test.ts:
```diff
@@ -0,0 +1,844 @@
+            })
+        })
+
+        describe("Set conversion rate", () => {
```
Add `Fail for non-owner`

ITStar10 replied:
Added

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getDIDRegistryAddress() external virtual view override returns(address) {
```
Why is this needed? I would like to avoid this if at all possible.

ITStar10 replied:
The `XPReward` contract is initialized with `Token` contract and `DIDRegistry` contract.
In some cases, we need to check the addresses of the `Token` and `DIDRegistry` contracts.

tahpot replied:
As discussed, this will be removed

tahpot commented on VDA-XP-Reward/contracts/VDAXPReward.sol:
```diff
@@ -0,0 +1,277 @@
+        if (conversionRate == 0) {
+            revert InvalidConversionRate();
+        }
+        if (!didRegistryContract.isRegistered(didAddress)) {
```
Let's remove this.

In the future we will support DID's registered on other blockchains, so this will break.

ITStar10 replied:
Removed.

tahpot reviewed (changes requested):
I have done an initial code review.

ITStar10 committed:
[vda-xp-reward] Update on feedback

ITStar10 committed:
[VDA-XP-Reward] Initial version

- Removed codes and test codes related to the `DIDRegistry` contract

ITStar10 committed:
[VDA-XP-Reward] Initial version

- Updated the denominator and conversionRate from `uint32` to `uint`
- Update deployment script : `scripts/deploy.ts`

ITStar10 committed:
[VDA-XP-Reward] Initial version

- Updated the hardhat config for verification

tahpot commented:
@ITStar10 There are still unresolved comments.

ITStar10 committed:
[VDA-XP-Reward] v0.1.0

- Updated verification of signature by using the `verifyData()`
- Updated proof time validation logic to check January correct
- Updated the test code

ITStar10 committed:
Merge branch 'develop' into feature/152-xp-reward-contract

ITStar10 committed:
[VDA-XP-Reward] v1.0.0

- Added comments in the test script

ITStar10 committed:
[VDA-XP-Reward] Initial version

- Separated the test case for failed in Jan

tahpot commented on VDA-XP-Reward/test/index.test.ts:
```diff
@@ -0,0 +1,920 @@
+        
+                })
+
+                it("Check in January", async () => {
```
Please add a test to check that current blockchain is Jan, but claim was issued in December -- works correctly.

ITStar10 replied:
Pls check the line#876:
![image](https://github.com/user-attachments/assets/9958a23b-9864-4b13-a029-9c1f330d0f0f)

ITStar10 replied:
Resolved
